### ISSUE TITLE
What is the official way to access session user in resolvers? Seeing inconsistent behaviors.

### ISSUE THREAD

**Issue by evanheckert** [nrbnlulu/strawberry-django-auth#578]
Sometimes the user exists in `info.context.request.user`, but sometimes that fails, but using `django.contrib.auth.get_user(info)` returns a user. Sometimes neither work, despite having a valid JWT well within expiration.

I do have a custom User model:

```python
# people/models.py
import uuid_utils.compat as uuid
from django.contrib.auth.models import AbstractUser
from django.db import models

class CustomUser(AbstractUser):
    id = models.UUIDField(primary_key=True, default=uuid.uuid7, editable=False)
    role = models.CharField(max_length=100, default="crew")
    full_name = models.CharField(max_length=255, editable=False)
    is_salesman = models.BooleanField(default=False)
    is_superuser = models.BooleanField(default=False)

    def __str__(self):
        return self.full_name
```
Here's my deps:
```
anyascii==0.3.2
asgiref==3.8.1
attrs==24.2.0
autobahn==24.4.2
automat==24.8.1
beautifulsoup4==4.12.3
certifi==2024.8.30
cffi==1.17.1
charset-normalizer==3.3.2
constantly==23.10.4
cryptography==43.0.1
daphne==4.1.2
defusedxml==0.7.1
django==5.0.9
django-cors-headers==4.4.0
django-filter==24.3
django-modelcluster==6.3
django-permissionedforms==0.1
django-stubs==4.2.7
django-stubs-ext==5.0.4
django-taggit==5.0.1
django-treebeard==4.7.1
django-types==0.19.1
django-unfold==0.37.0
djangorestframework==3.15.2
draftjs-exporter==5.0.0
et-xmlfile==1.1.0
filetype==1.2.0
graphql-core==3.2.4
hyperlink==21.0.0
idna==3.8
incremental==24.7.2
l18n==2021.3
laces==0.1.1
mypy==1.7.1
mypy-extensions==1.0.0
openpyxl==3.1.5
pillow==10.4.0
pillow-heif==0.18.0
psycopg2-binary==2.9.9
pyasn1==0.6.0
pyasn1-modules==0.4.0
pycparser==2.22
pyjwt==2.9.0
pyopenssl==24.2.1
python-dateutil==2.9.0.post0
python-decouple==3.8
pytz==2024.1
requests==2.32.3
ruff==0.6.4
sentry-sdk==2.13.0
service-identity==24.1.0
setuptools==74.1.2
six==1.16.0
soupsieve==2.6
sqlparse==0.5.1
strawberry-django-auth==0.377.0
strawberry-graphql==0.239.2
strawberry-graphql-django==0.47.2
telepath==0.3.1
twisted==24.7.0
txaio==23.1.1
types-psycopg2==2.9.21.20240819
types-pytz==2024.1.0.20240417
types-pyyaml==6.0.12.20240808
typing-extensions==4.12.2
urllib3==2.2.2
uuid-utils==0.9.0
wagtail==6.2.1
wagtail-modeladmin==2.0.0
willow==1.8.0
zope-interface==7.0.3
```

**Comment by nrbnlulu** [comment]
Hey, I can't really help with "sometimes" statements. Can you try to create an MRE or some instructions for the cases you said?

**Comment by evanheckert** [comment]
Thanks for the reply! And yes I should have been much more clear in my original post. 

I can debug more easily (and track down a reproducible code example) if I know what the intended behavior for accessing the session user in a resolver? I didn't see it in the documentation. That seems odd to me for a package focused on auth, so it's likely it's there and I missed it.

**Comment by nrbnlulu** [comment]
Sorry, I can't understand what you said :disappointed: can you rephrase?

**Comment by evanheckert** [comment]
No problem!

**What is the recommended way to access the session user in a resolver?**

I’m seeing inconsistent behavior when trying to access the user in my resolvers. Sometimes I can access the user with info.context.request.user, but other times this does not work, and I can only retrieve the user using django.contrib.auth.get_user(info). In some cases, neither method works, even though the JWT is valid and not expired.

Could you clarify the intended way to access the session user within a resolver? I wasn’t able to find this information in the documentation, so I’m unsure whether I’m missing something or if this behavior is unexpected.

**Comment by nrbnlulu** [comment]
Thats kinda wierd behaviour can you show your setup (asgi/wsgi and dajgno settings)

**Comment by evanheckert** [comment]
```python
# settings.py with irrelevant stuff removed
from datetime import timedelta

from gqlauth.settings_type import GqlAuthSettings
import os.path
from pathlib import Path

import django_stubs_ext
from corsheaders.defaults import default_headers
from decouple import config

django_stubs_ext.monkeypatch()

BASE_DIR = Path(__file__).resolve().parent.parent.parent

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "corsheaders.middleware.CorsMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "gqlauth.core.middlewares.django_jwt_middleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "wagtail.contrib.redirects.middleware.RedirectMiddleware",
]

ROOT_URLCONF = "thermal_admin.urls"

WSGI_APPLICATION = "thermal_admin.wsgi.application"

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": DB_NAME,
        "USER": DB_USER,
        "PASSWORD": DB_PASS,
        "HOST": "localhost",
        "PORT": "",
    }
}

AUTH_USER_MODEL = "people.CustomUser"  # new
ALLOW_LOGIN_NOT_VERIFIED = True
LOGIN_REQUIRE_CAPTCHA = False

AUTHENTICATION_BACKENDS = [
    "django.contrib.auth.backends.ModelBackend",
]

GQL_AUTH = GqlAuthSettings(
    LOGIN_REQUIRE_CAPTCHA=False,
    REGISTER_REQUIRE_CAPTCHA=False,
    JWT_EXPIRATION_DELTA=timedelta(minutes=240),
    JWT_REFRESH_EXPIRATION_DELTA=timedelta(days=7),
    JWT_LONG_RUNNING_REFRESH_TOKEN=True,
)

# Password validation
# https://docs.djangoproject.com/en/5.0/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]
```

```python
# asgi.py
import os

from django.core.asgi import get_asgi_application

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'thermal_admin.settings')

application = get_asgi_application()
```

**Comment by nrbnlulu** [comment]
and your schema please. sorry :disappointed:

**Comment by kubami** [comment]
Do any of those details matter? I think the question is quite clear "What is the official way to access session user in resolvers?". 
Do @evanheckert settings / schemas / asgi setup matter? Shouldn't there be a one way?

The docs usually mention: _we just provide you a user in `info.context.request.`_

But, as with `strawberry-django` sometimes the user, depending on the type of request is in a different place.
`strawberry-django` has this method:

```python
def get_current_user(info: Info, *, strict: bool = False) -> UserType:
    """Get and return the current user based on various scenarios."""
    request = get_request(info)

    try:
        user = request.user
    except AttributeError:
        try:
            # queries/mutations in ASGI move the user into consumer scope
            user = request.consumer.scope["user"]  # type: ignore
        except AttributeError:
            # websockets / subscriptions move scope inside of the request
            user = request.scope.get("user")  # type: ignore

    if user is None:
        raise ValueError("No user found in the current request")

    # Access an attribute inside the user object to force loading it in async contexts.
    _ = user.is_authenticated

    return user
```

**Comment by nrbnlulu** [comment]
should be in `context.request.user` if it supports `setattr` otherwise its `context.request["user"]`
